### PR TITLE
[FLINK-28219] Refactor TableStoreCatalog: Introduce a dedicated Catalog for table store

### PR DESCRIPTION
**Pull request by JingsongLi**
We currently have developed a Flink's Catalog.
If we expose this Catalog directly to other connector developers, it is not good and there will be many unsupported interfaces and capabilities.

So here we create a tablestore dedicated catalog.

### CODE REVIEW

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+import org.apache.flink.table.store.file.schema.TableSchema;
+import org.apache.flink.table.store.file.schema.UpdateSchema;
+
+import java.util.List;
+
+/**
+ * This interface is responsible for reading and writing metadata such as database/table from a
+ * table store catalog.
+ */
+public interface Catalog extends AutoCloseable {
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java:32]: Nit: add `@PublicEvolving` annotation?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+    /**
+     * Get names of all tables under this database. An empty list is returned if none exists.
+     *
+     * @return a list of the names of all tables in this database
+     */
+    List<String> listTables(String databaseName);
+
+    /**
+     * Returns the table location path identified by the given {@link ObjectPath}.
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java:60]: Nit: Return

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+     *
+     * @param tablePath Path of the table
+     * @return The requested table location
+     */
+    Path getTableLocation(ObjectPath tablePath);
+
+    /**
+     * Returns a {@link TableSchema} identified by the given {@link ObjectPath}.
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java:68]: Ditto

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+     *
+     * @param tablePath Path of the table to be dropped
+     * @param ignoreIfNotExists Flag to specify behavior when the table does not exist: if set to
+     *     false, throw an exception, if set to true, do nothing.
+     * @throws TableNotExistException if the table does not exist
+     */
+    void dropTable(ObjectPath tablePath, boolean ignoreIfNotExists) throws TableNotExistException;
+
+    /**
+     * Creates a new table.
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java:95]: Ditto

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+     * @param tableSchema the table definition
+     * @param ignoreIfExists flag to specify behavior when a table already exists at the given path:
+     *     if set to false, it throws a TableAlreadyExistException, if set to true, do nothing.
+     * @throws TableAlreadyExistException if table already exists and ignoreIfExists is false
+     */
+    void createTable(ObjectPath tablePath, UpdateSchema tableSchema, boolean ignoreIfExists)
+            throws TableAlreadyExistException;
+
+    /**
+     * Modifies an existing table.
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java:107]: Ditto

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    List<String> listDatabases();
+
+    /**
+     * Drop a database.
+     *
+     * @param name Name of the database to be dropped.
+     * @param cascade Flag to specify behavior when the database contains table or function: if set
+     *     to true, delete all tables and functions in the database and then delete the database, if
+     *     set to false, throw an exception.
+     * @throws DatabaseNotEmptyException if the given database is not empty and isRestrict is true
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java:48]: isRestrict?

Comment (LadyForest) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java:48]: Missing a param like `ignoreIfNotExists`?

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/FlinkCatalog.java [hunk 7]
```diff
@@ -16,39 +16,192 @@
...
+
+        return UpdateSchema.fromCatalogTable(table);
     }
 
     // --------------------- unsupported methods ----------------------------
 
+    @Override
+    public void createDatabase(String name, CatalogDatabase database, boolean ignoreIfExists)
+            throws CatalogException {
+        throw new UnsupportedOperationException();
```
Comment (LadyForest) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/FlinkCatalog.java:191]: just wondered why

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/FlinkCatalog.java [hunk 8]
```diff
@@ -16,39 +16,192 @@
...
+    }
+
+    @Override
+    public List<String> listDatabases() throws CatalogException {
+        return catalog.listDatabases();
+    }
+
+    @Override
+    public boolean databaseExists(String databaseName) throws CatalogException {
+        return true;
```
Comment (LadyForest) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/FlinkCatalog.java:85]: always `true`?

File: flink-table-store-connector/src/test/java/org/apache/flink/table/store/connector/FlinkCatalogTest.java [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+        List<String> tables = catalog.listTables("db1");
+        Assert.assertEquals(1L, tables.size());
+        Assert.assertEquals(this.path1.getObjectName(), tables.get(0));
+    }
+
+    @Test
+    public void testDropDb_DatabaseNotEmptyException() throws Exception {
+        catalog.createTable(this.path1, this.createTable(), false);
+        this.exception.expect(DatabaseNotEmptyException.class);
+        this.exception.expectMessage("Database db1 in catalog test-catalog is not empty");
```
Comment (LadyForest) [inline flink-table-store-connector/src/test/java/org/apache/flink/table/store/connector/FlinkCatalogTest.java:255]: mixed-use of two assertions?

Comment (LadyForest) [inline flink-table-store-connector/src/test/java/org/apache/flink/table/store/connector/FlinkCatalogTest.java:255]: ditto

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java (hunk 1 above)
Comment (JingsongLi) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java:32]: We don't public any interfaces now.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java (hunk 6 above)
Comment (JingsongLi) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java:48]: I was previously considering database creation all the way to lazy, but exposing it is more comprehensive.

**Comment by JingsongLi** [comment]
> Q1: What are the criteria for choosing which interface not to implement. E.g. createDatabase is unsupported, which is confusing to me.

I was previously considering database creation all the way to lazy, but exposing it is more comprehensive.
So yes, I will change this.

> O2: It seems like FlinkCatalog serves as a wrapper to interact with Flink. Does it only accept the 3-rd party ecosystem's database as the inner impl?

See above comment, we should change back to support the previous functionality

**Review by LadyForest — APPROVED** [review]
Thanks for the explanation and update, looks good to me :)

**Pull request merged**